controllers/nstemplateset/space_roles.go

```diff
@@ -0,0 +1,136 @@
+package nstemplateset
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	applycl "github.com/codeready-toolchain/toolchain-common/pkg/client"
+	"github.com/pkg/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+)
+
+type spaceRolesManager struct {
+	*statusManager
+}
+
+// ensure ensures that the space roles for the users exist.
+// Returns `true, nil` if something was changed, `false, nil` if nothing changed, `false, err` if an error occurred
+// TODO: support update/removal/outdated
+func (r *spaceRolesManager) ensure(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
+	logger = logger.WithValues("nstemplateset_name", nsTmplSet.Name)
+
+	nss, err := fetchNamespacesByWorkspace(r.Client, nsTmplSet.Name)
+	if err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
+			"failed to list namespaces for workspace '%s'", nsTmplSet.Name)
+	}
+	logger.Info("ensuring space roles", "namespace_count", len(nss), "role_count", len(nsTmplSet.Spec.SpaceRoles))
+	changed := false
+	for _, ns := range nss {
+		// space roles to apply now
+		lastAppliedSpaceRoleObjs, err := r.getLastAppliedSpaceRoles(&ns)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve last applied space roles")
+		}
+		spaceRoleObjs, err := r.getSpaceRoles(&ns, nsTmplSet.Spec.SpaceRoles)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve space roles to apply")
+		}
+
+		c, err := deleteOutdatedObjects(logger, r.Client, lastAppliedSpaceRoleObjs, spaceRoleObjs)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusUpdateFailed, err, "failed to delete redundant objects in namespace '%s'", ns.Name)
+		}
+		changed = changed || c
+		// labels to apply on all new objects
+		var labels = map[string]string{
+			toolchainv1alpha1.ProviderLabelKey: toolchainv1alpha1.ProviderLabelValue,
+			toolchainv1alpha1.OwnerLabelKey:    nsTmplSet.GetName(), // TODO: is it needed?
```

Comment:
```
// TODO: is it needed?
```
yes, we need the owner label there so the mapper can map Role & RoleBindings to the name of the NSTemplateSet.

We could technically add also the `username` label and `spacerole` label so we could easily list them, but since we are not sure if we would need them, then let's keep it as it is for now (and introduces them later when/if needed.

Reply:
thanks for your response @MatousJobanek! I was hoping you'd notice this `//TODO` comment and provide me with an answer during the review 😛

Reply:
comment removed in e5d2100